### ISSUE TITLE
[Bug]: Segmentation Fault CALL _timescaledb_internal.policy_compression()

### ISSUE THREAD

**Issue by aarondglover** [timescale/timescaledb#6186]
### What type of bug is this?

Crash, Unexpected error

### What subsystems and features are affected?

Compression

### What happened?

There is a bit of a back story to my system. It is currently hosted on Timescale Managed Service (MST) 

The system was upgraded from pg13 to pg15 on the 5th October.

I now have a problem with compression - specifically my server crashes (segmentation fault) when attempting to run the compression job.

- I have a single hypertable in this database.
- I initially enabled compression on the table as follows;
```
 ALTER TABLE meter_data SET (
  timescaledb.compress,
  timescaledb.compress_segmentby = 'meter_id, meter_channel_id',
  timescaledb.compress_orderby = 'offset_timestamp DESC'
);

Not sure if relevant but (meter_id is a varchar, meter_channel_id is an int/bigint)
```
- I created a policy for 52 weeks `SELECT add_compression_policy('meter_data', INTERVAL '52 weeks');` This worked perfectly. The table compressed and I could update/delete/insert data into the compressed chunks etc. All working perfectly at this point.
- I removed the policy and added a new policy to compress older than 26 weeks. The compression seems to run ok.
- I then removed the policy and added a new policy to compress on 52 weeks (back to original setting)... It seemed to have applied ok, but i now get an error/crash when the compression job attempts to run
```
[mdm-64] 2023-10-10T05:47:21.325147 [postgresql-15] [229-1] pid=1965095,user=,db=,app=,client= LOG: server process (PID 2331541) was terminated by signal 11: Segmentation fault
[mdm-64] 2023-10-10T05:47:21.325456 [postgresql-15] [229-2] pid=1965095,user=,db=,app=,client= DETAIL: Failed process was running: CALL _timescaledb_internal.policy_compression()
[mdm-64] 2023-10-10T05:47:21.325490 [postgresql-15] [230-1] pid=1965095,user=,db=,app=,client= LOG: terminating any other active server processes
[mdm-64] 2023-10-10T05:47:21.331173 [postgresql-15] [231-1] pid=1965095,user=,db=,app=,client= LOG: all server processes terminated; reinitializing
[mdm-64] 2023-10-10T05:47:21.718707 [postgresql-15] [232-1] pid=2331583,user=,db=,app=,client= LOG: database system was interrupted; last known up at 2023-10-10 05:47:04 GMT
```

Output when running at PSQL:
```
CALL run_job(1119)
bind <unnamed> to <unnamed>
Executing policy_compression with parameters {"hypertable_id": 14, "compress_after": "364 days"}
rehashing catalog cache id 54 for pg_class; 257 tups, 128 buckets
```

Timescale support are trying to capture a core-dump

### TimescaleDB version affected

2.11.2

### PostgreSQL version used

15.4

### What operating system did you use?

Unknown (AIVEN TSDB MST)

### What installation method did you use?

Other

### What platform did you run on?

Managed Service for TimescaleDB (MST/Aiven)

### Relevant log output and stack trace

```bash
[mdm-64] 2023-10-10T05:47:21.325147 [postgresql-15] [229-1] pid=1965095,user=,db=,app=,client= LOG: server process (PID 2331541) was terminated by signal 11: Segmentation fault
[mdm-64] 2023-10-10T05:47:21.325456 [postgresql-15] [229-2] pid=1965095,user=,db=,app=,client= DETAIL: Failed process was running: CALL _timescaledb_internal.policy_compression()
[mdm-64] 2023-10-10T05:47:21.325490 [postgresql-15] [230-1] pid=1965095,user=,db=,app=,client= LOG: terminating any other active server processes
[mdm-64] 2023-10-10T05:47:21.331173 [postgresql-15] [231-1] pid=1965095,user=,db=,app=,client= LOG: all server processes terminated; reinitializing
[mdm-64] 2023-10-10T05:47:21.718707 [postgresql-15] [232-1] pid=2331583,user=,db=,app=,client= LOG: database system was interrupted; last known up at 2023-10-10 05:47:04 GMT

Output when running at PSQL:

CALL run_job(1119);
bind <unnamed> to <unnamed>
Executing policy_compression with parameters {"hypertable_id": 14, "compress_after": "364 days"}
rehashing catalog cache id 54 for pg_class; 257 tups, 128 buckets
```

### How can we reproduce the bug?

```bash
This is consistently reproducible on my system by attempting to run the compression job

ie:  CALL  run_job(1119);
```

**Comment by aarondglover** [comment]
interesting comment noted in other bug for 2.12.0 that compression setting changes are blocked if there are compressed blocks. Does this comment only apply to the compression settings? ie: `compress_segmentby  `and `compress_orderby ` or does this comment refer to changes of the policy such as 52 weeks to 26 weeks etc?

https://github.com/timescale/timescaledb/issues/6137#issuecomment-1744382399

**Comment by konskov** [comment]
Hi @aarondglover, thank you for reaching out. 
Have you tried manually compressing or recompressing chunks older than 52 weeks and did you get the segmentation fault?
Also, it would be most helpful if there was a stacktrace available. Since this is MST I’d suggest communicating with support to get one from the Aiven platform.
Regarding the question, that comment refers to the compression settings (orderby, segmentby). Making changes to the compression policy is not blocked, you can do so through the `alter_job` API.

**Comment by konskov** [comment]
as a side note, and most probably unrelated to this issue, using text instead of varchar is a better practice

**Comment by aarondglover** [comment]
> Hi @aarondglover, thank you for reaching out. Have you tried manually compressing or recompressing chunks older than 52 weeks and did you get the segmentation fault? Also, it would be most helpful if there was a stacktrace available. Since this is MST I’d suggest communicating with support to get one from the Aiven platform. Regarding the question, that comment refers to the compression settings (orderby, segmentby). Making changes to the compression policy is not blocked, you can do so through the `alter_job` API.

Hi @konskov I have tried, decompress_chunk and recompress. Both trigger a segmentation fault too.

I have also tried re-indexing the chunk prior to the above methods which worked fine, but did not result in decompress_chunk working.

I have a ticket open with Timescale and they have captured some core-dumps but I don't have a view to those.

**Comment by aarondglover** [comment]
> Hi @aarondglover, thank you for reaching out. Have you tried manually compressing or recompressing chunks older than 52 weeks and did you get the segmentation fault? Also, it would be most helpful if there was a stacktrace available. Since this is MST I’d suggest communicating with support to get one from the Aiven platform. Regarding the question, that comment refers to the compression settings (orderby, segmentby). Making changes to the compression policy is not blocked, you can do so through the `alter_job` API.

Hi @konskov I have tried, decompress_chunk and recompress. Both actually trigger a segmentation fault too.

**Comment by konskov** [comment]
Updating this issue for visibility: the segfault is due to our handling of expression indexes

**Comment by aarondglover** [comment]
Is this an issue in 2.12 too?

**Comment by konskov** [comment]
Hello @aarondglover - yes, unfortunately this was introduced in 2.11.0 (2.10.3 did not have this issue) and is also reproducible in our main branch. But we are currently working on a fix.

**Comment by aarondglover** [comment]
Hi @konskov I note your comment in your recent commit...

_This patch fixes this by simply ignoring the expression columns in the index, since we cannot currently segment by an expression._

I currently DO segment by a field that has an expression index on it... Does this mean I will NOT be able to enable compression on my hypertable if I have segment by a field that has an expression index on it? Ie... Lower()/upper() etc

**Comment by konskov** [comment]
hi @aarondglover, what that comment meant by "ignoring" was that warnings like `WARNING: column "col1" should be used for segmenting or ordering` will not be issued for the expression columns - though perhaps it might be better to raise a warning in these cases too. Since we cannot currently segment by an expression - meaning that we cannot specify `timescaledb.compress_segmentby = 'lower(col1)'` but instead `timescaledb.compress_segmentby = 'col1'` .
You would of course still be able to enable compression on the hypertable with an expression index defined. The comment was just about printing the warning. Hope this makes things clearer!

**Comment by konskov** [comment]
I believe that #6222 which was merged recently fixes this issue, so I am closing it. It will most probably be part of the next release (after 2.12.2). @aarondglover could you please check once it is released that it solves the problem and let us know?

**Comment by aarondglover** [comment]
@konskov I am waiting very keenly for this release. I use TSDB in Managed Service - so I am hoping they release into this environment very soon. It seems there is often some delay to deploy new version into TSDB Managed Service (AIVEN) though :(

#6222 isn't mentioned in CHANGELOG.md

Also confused about CHANGELOG.md because it mentions 2.12.2 release on 19th October. But the docs website makes no mention of this release. Also if 2.12.2 was released on that date. How could my issue be solved in 2.12.2? Was it backported to that release (note: the "not-backported" label on #6222)

Before I go raising an issue and/or PR on the docs repository. I wonder if you can clarify the situation regarding release 2.12.2

**Comment by konskov** [comment]
hi @aarondglover ! Thank you for your patience with this issue.

I expect #6222 to be part of the next release, which as mentioned will be the one _**after**_ 2.12.2. I did not explicitly name it because I didn’t know if it was going to be 2.12.3 or 2.13.0. But in any case, release 2.12.2 was tagged before the bugfix was merged so it is not going to be part of 2.12.2. Hope this clarifies the situation a bit.
The not-backported label doesn’t mean it should be included in the past release, it is added as part of a workflow we have in place to facilitate preparing future releases. 

Regarding the release notes in our documentation, you are correct in noting they are not up-to-date. We will soon update them to the latest release, which as of now is 2.12.2.
